inductor codecache: include private inductor configs in cache key

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/torchtitan/issues/1185

It looks like inductor's logic to include inductor configs in the cache key skips configs with a leading underscore by default. This came up in torchtitan - there's an asyncTP pipelining pass in inductor gated by a private config, and by not caching on the config we were attempting to use asyncTP when we shouldn't be.

I'm not sure how worried we should be on the blast radius of this change. On the one hand:

(1) it technically fixes any silent correctness issues in the cache around any other private inductor configs (it looks like there are a few)

(2) there is some risk that there are some "harmless" configs that we are now including in the key, which may increase false negatives. I do see that there is an explicit list for "configs we want to ignore for caching" (`_save_config_ignore`), so my hope is that all harmless configs are already encapsulated there.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153672



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

bdhirsh committed:
inductor codecache: include private inductor configs in cache key

[ghstack-poisoned]

oulgen reviewed (commented):
I'm OK with this but before committing it can you do an audit of what was not being included previous and is now included?

you might need to prematurely include these in the ignore list in case they contain file paths etc

bdhirsh commented:
Here's the diff from tlparse: https://www.internalfb.com/intern/diffing/?paste_number=1812905517

I don't see any hardcoded paths. The only ones that seem risky to me are the private configs that correspond to custom passes:
```
_fuse_ddp_communication_passes
_pre_fusion_custom_pass
```

I'll add them to the list of configs to ignore for now to reduce blast radius, although we may want to revisit hashing of the custom passes eventually :)

bdhirsh committed:
Update on "inductor codecache: include private inductor configs in cache key"

Fixes https://github.com/pytorch/torchtitan/issues/1185

It looks like inductor's logic to include inductor configs in the cache key skips configs with a leading underscore by default. This came up in torchtitan - there's an asyncTP pipelining pass in inductor gated by a private config, and by not caching on the config we were attempting to use asyncTP when we shouldn't be.

I'm not sure how worried we should be on the blast radius of this change. On the one hand:

(1) it technically fixes any silent correctness issues in the cache around any other private inductor configs (it looks like there are a few)

(2) there is some risk that there are some "harmless" configs that we are now including in the key, which may increase false negatives. I do see that there is an explicit list for "configs we want to ignore for caching" (`_save_config_ignore`), so my hope is that all harmless configs are already encapsulated there.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen commented on torch/_inductor/config.py:
```diff
@@ -1569,6 +1569,8 @@ class trace:
+    "_pre_fusion_custom_pass",
 ]
 
 _cache_config_ignore_prefix: list[str] = [
```
You need to add them here, and then do the same thing in caching code for what we do with post_grad_custom_post_pass

Skylion007 commented on torch/utils/_config_module.py:
```diff
@@ -508,9 +508,13 @@ def save_config(self) -> bytes:
         )
 
-    def save_config_portable(self) -> dict[str, Any]:
+    def save_config_portable(
```
Typing is in a different .pyi module, might want to merge them together now...

bdhirsh replied:
Oh huh, I didn't realize we were putting the typing in a separate .pyi file. It looks like the typing is at least partially duplicated across this file and `torch/utils/_config_typing.pyi`?

Merging sounds reasonable but I don't have all of the context... for now I'm just going to keep this change less disruptive and mirror my new kwarg into the pyi file.

bdhirsh committed:
Update on "inductor codecache: include private inductor configs in cache key"

Fixes https://github.com/pytorch/torchtitan/issues/1185

It looks like inductor's logic to include inductor configs in the cache key skips configs with a leading underscore by default. This came up in torchtitan - there's an asyncTP pipelining pass in inductor gated by a private config, and by not caching on the config we were attempting to use asyncTP when we shouldn't be.

I'm not sure how worried we should be on the blast radius of this change. On the one hand:

(1) it technically fixes any silent correctness issues in the cache around any other private inductor configs (it looks like there are a few)

(2) there is some risk that there are some "harmless" configs that we are now including in the key, which may increase false negatives. I do see that there is an explicit list for "configs we want to ignore for caching" (`_save_config_ignore`), so my hope is that all harmless configs are already encapsulated there.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/codecache.py:
```diff
@@ -1330,6 +1330,14 @@ def _check_can_cache(gm: torch.fx.GraphModule) -> None:
+                raise BypassFxGraphCache("Unsupported _pre_fusion_custom_pass")
+        for p in config._fuse_ddp_communication_passes:
+            if callable(p) and not isinstance(p, CustomGraphPass):
+                raise BypassFxGraphCache("Unsupported _fuse_ddp_communication_pass")
```
One thing I did not explicitly fix in this PR is the fact that this config (`_fuse_ddp_communication_passes`) is a bit weird - it's typed as a `list[Union[Callable[..., None], str]]`, where:

(1) from an fbcode search, we only pass in strings today (for hardcoded passes inside inductor)

(2) Looking at at least one example custom pass, it's signature is `def fuse_ddp_with_concat_op(graph: fx.Graph, bucket_size_mb: int) -> None:` (code: https://github.com/pytorch/pytorch/blob/2ce0b66db8b6a22e90b430a73b8914c2d73512e9/torch/_inductor/fx_passes/ddp_fusion.py#L513C1-L513C75). I'm not sure how much flexibility distributed folks need/want in giving their custom passes additional arguments aside from the FX graph itself.

The `CustomGraphPass` today is hardcoded to expect passes with a function `def __call__(self, graph: torch.fx.graph.Graph) -> None`.

I don't have enough context on these distributed passes (maybe @wconstab or @fegin would know?), but for now I side-stepped trying to add more typing to the configs to allow `CustomGraphPass`. So the state of this PR is that if you want to use one of these custom pass configs out of tree and pass in a custom callable, we will cache miss for safety.

bdhirsh commented on test/inductor/test_codecache.py:
```diff
@@ -2046,6 +2078,93 @@ def test_hash_config_changes(self):
+                    x.getMessage()
+                    == "Bypassing FX Graph Cache because 'Unsupported _fuse_ddp_communication_pass'"
+                    for x in logs
+                )
```
@oulgen I added logic to explicitly test that we are raising a reasonable bypass error.

I was going to try to re-use some existing logging infra in the testing to capture codecache logs (e.g. this [function](https://github.com/pytorch/pytorch/blob/main/test/inductor/test_debug_trace.py#L43)), but I wasn't able to because it looks like the `torch._inductor.codecache` logs don't use `torch._logging.getArtifactLogger`.

I ended up rolling my own tiny util in this file for capturing the codecache logs to assert against, but lmk if you'd rather I do something else

oulgen replied:
> but I wasn't able to because it looks like the torch._inductor.codecache logs don't use torch._logging.getArtifactLogger.

would it be hard to migrate it over?

oulgen replied:
I'm equally clueless :) so feel free to try, as long as you dont break tlparse i think whatever you do is probably fine

bdhirsh replied:
done in a new PR here: https://github.com/pytorch/pytorch/pull/153766

bdhirsh committed:
Update on "inductor codecache: include private inductor configs in cache key"

Fixes https://github.com/pytorch/torchtitan/issues/1185

It looks like inductor's logic to include inductor configs in the cache key skips configs with a leading underscore by default. This came up in torchtitan - there's an asyncTP pipelining pass in inductor gated by a private config, and by not caching on the config we were attempting to use asyncTP when we shouldn't be.

I'm not sure how worried we should be on the blast radius of this change. On the one hand:

(1) it technically fixes any silent correctness issues in the cache around any other private inductor configs (it looks like there are a few)

(2) there is some risk that there are some "harmless" configs that we are now including in the key, which may increase false negatives. I do see that there is an explicit list for "configs we want to ignore for caching" (`_save_config_ignore`), so my hope is that all harmless configs are already encapsulated there.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/codecache.py:
```diff
@@ -855,14 +855,44 @@ def __init__(
+    # - if any of them are set to custom callables, we will need to cache miss
+    # Future work is for someone to find any places where these functions are used
+    # and force them to be of type CustomGraphPass, so we can guarantee serialization.
+    def _get_custom_pass_detail_unsafe(self, custom_pass: Any) -> Optional[Any]:
```
@oulgen this is the main ugly code that I added to deal with the fact that inductor has some existing custom pass configs that can be a few different types (strings or callables)

bdhirsh committed:
Update on "inductor codecache: include private inductor configs in cache key"

Fixes https://github.com/pytorch/torchtitan/issues/1185

It looks like inductor's logic to include inductor configs in the cache key skips configs with a leading underscore by default. This came up in torchtitan - there's an asyncTP pipelining pass in inductor gated by a private config, and by not caching on the config we were attempting to use asyncTP when we shouldn't be.

I'm not sure how worried we should be on the blast radius of this change. On the one hand:

(1) it technically fixes any silent correctness issues in the cache around any other private inductor configs (it looks like there are a few)

(2) there is some risk that there are some "harmless" configs that we are now including in the key, which may increase false negatives. I do see that there is an explicit list for "configs we want to ignore for caching" (`_save_config_ignore`), so my hope is that all harmless configs are already encapsulated there.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

malfet commented:
@pytorchbot revert -m "Looks like it regressed pr_time_benchmarks, see https://hud.pytorch.org/hud/pytorch/pytorch/ba3f91af97c2453b72c4876c69b47ba62efc7e85/1?per_page=50&name_filter=pr_time_benchmarks&mergeEphemeralLF=true" -c weird

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update

[ghstack-poisoned]

bdhirsh committed:
Update on "inductor codecache: include private inductor configs in cache key"

Fixes https://github.com/pytorch/torchtitan/issues/1185

It looks like inductor's logic to include inductor configs in the cache key skips configs with a leading underscore by default. This came up in torchtitan - there's an asyncTP pipelining pass in inductor gated by a private config, and by not caching on the config we were attempting to use asyncTP when we shouldn't be.

I'm not sure how worried we should be on the blast radius of this change. On the one hand:

(1) it technically fixes any silent correctness issues in the cache around any other private inductor configs (it looks like there are a few)

(2) there is some risk that there are some "harmless" configs that we are now including in the key, which may increase false negatives. I do see that there is an explicit list for "configs we want to ignore for caching" (`_save_config_ignore`), so my hope is that all harmless configs are already encapsulated there.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge -r

bdhirsh committed:
Update

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -r

bdhirsh committed:
Update

[ghstack-poisoned]
